File: 2458 Height of Binary Tree After Subtree Removal Queries/2458.py
class Solution:
    def treeQueries(self, root: Optional[TreeNode], queries: List[int]) -> List[int]:
        removed = [0]*(10**5+1)

        def dfs(cur, h):
            if not cur: return

            nonlocal max_h, left_first

            removed[cur.val] = max(removed[cur.val], max_h)
            max_h = max(max_h, h)

            dfs(cur.left if left_first else cur.right, h+1)
            dfs(cur.right if left_first else cur.left, h+1)

        for b in [True, False]:
            max_h = 0
            left_first = b
            dfs(root, 0)

        ans = []

        for q in queries:
            ans.append(removed[q])

        return ans

File: 2458 Height of Binary Tree After Subtree Removal Queries/2458.cpp
class Solution {
public:
    int maxH;
    bool leftFirst;
    int removed[100001];
    
    void dfs(TreeNode* cur, int h){
        if (cur == nullptr) return;

        removed[cur->val] = max(removed[cur->val], maxH);
        maxH = max(maxH, h);

        dfs(leftFirst ? cur->left : cur->right, h+1);
        dfs(leftFirst ? cur->right : cur->left, h+1);
    }

    vector<int> treeQueries(TreeNode* root, vector<int>& queries) {
        leftFirst = true;
        dfs(root, 0);

        maxH = 0;
        leftFirst = false;
        dfs(root, 0);

        vector<int> ans;

        for (int q: queries) ans.push_back(removed[q]);

        return ans;
    }
};
